src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -55,26 +57,42 @@ protected override object VisitConstructor(ConstructorCallSite constructorCallSi
 #endif
         }
 
-        protected override object VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
+        protected override object VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
         {
-            return VisitCache(singletonCallSite, context, context.Scope.Engine.Root, RuntimeResolverLock.Root);
+            var lockType = RuntimeResolverLock.Root;
+            bool lockTaken = false;
+
+            if ((context.AcquiredLocks & lockType) == 0)
```

Comment:
@pakrym notice that this check here is problematic and can no longer be used since we're taking a unique lock per singleton. We don't want to skip it here anymore.

Reply:
I see. I'll fix both comments with test. Thanks.